pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java

```diff
@@ -193,69 +191,74 @@ public NettyChannelInboundHandler(RequestHandler handler, NettyServerMetrics met
       this(handler, metric, 100);
     }
 
-    public enum State {
-      INIT,
-      REQUEST_RECEIVED,
-      RESPONSE_WRITTEN,
-      RESPONSE_SENT,
-      EXCEPTION
-    }
-
-    /**
-     * Server Channel Handler State
-     */
-    private State _state = State.INIT;
-
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) {
-      _requestStartTime = System.currentTimeMillis();
+      final long requestStartTime = System.currentTimeMillis();
       LOGGER.debug("Request received by server !!");
-      _state = State.REQUEST_RECEIVED;
-      ByteBuf request = (ByteBuf) msg;
-      _lastRequsetSizeInBytes = request.readableBytes();
+
+      final ByteBuf request = (ByteBuf) msg;
+      final long requestSizeInBytes = request.readableBytes();
 
       //Call processing handler
-      _lastProcessingLatency = MetricsHelper.startTimer();
-      byte[] response = _handler.processRequest(ctx, request);
-      _lastProcessingLatency.stop();
-
-      // Send Response
-      ByteBuf responseBuf = Unpooled.wrappedBuffer(response);
-      _lastSendResponseLatency = MetricsHelper.startTimer();
-      ChannelFuture f = ctx.writeAndFlush(responseBuf);
-      _state = State.RESPONSE_WRITTEN;
-      f.addListener(this);
-      request.release();
+      final TimerContext requestProcessingLatency = MetricsHelper.startTimer();
+      final ChannelHandlerContext requestChannelHandlerContext = ctx;
+      ListenableFuture<byte[]> serializedQueryResponse = _handler.processRequest(ctx, request);
+      Futures.addCallback(serializedQueryResponse, new FutureCallback<byte[]>() {
```

Comment:
It will be immediately executed on the same thread. Executing on the same thread can be dangerous, say, for UI threads or "main" server thread which can make the system non-responsive. In our case, it will be executed on the queryRunner which is what we want